tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSGen.scala

```diff
@@ -237,16 +263,36 @@ private[emitter] final class JSGen(val semantics: Semantics,
       if (containsOnlyValidChars()) "$i_" + module
       else buildValidName()
 
-    VarRef(Ident(varName, Some(module)))
+    VarRef(Ident(avoidClashWithGlobalRef(varName), Some(module)))
   }
 
   def envField(field: String, subField: String, origName: Option[String] = None)(
       implicit pos: Position): VarRef = {
-    VarRef(Ident("$" + field + "_" + subField, origName))
+    VarRef(Ident(avoidClashWithGlobalRef("$" + field + "_" + subField),
+        origName))
   }
 
   def envField(field: String)(implicit pos: Position): VarRef =
-    VarRef(Ident("$" + field))
+    VarRef(Ident(avoidClashWithGlobalRef("$" + field)))
+
+  def avoidClashWithGlobalRef(envFieldName: String): String = {
+    @tailrec
+    def slowPath(lastNameTried: String): String = {
+      val nextNameToTry = "$" + lastNameTried
+      if (mentionedDangerousGlobalRefs.contains(nextNameToTry))
+        slowPath(nextNameToTry)
+      else
+        nextNameToTry
+    }
```

Comment:
Yes. It should happen virtually never anyway. And the likelihood that it takes more than 1 iteration is so close to 0 that a cache really doesn't make any sense. It would slow things down.